### DISCUSSION TITLE
The error will be redacted in production builds

### DISCUSSION THREAD

**Discussion by yrnana** [TanStack/query#8286] [Q&A]
I'm using nextjs 14 app router and tanstack query 5.40+ (no persist)

https://tanstack.com/query/v5/docs/framework/react/guides/advanced-ssr#streaming-with-server-components
I'm trying to use no-await prefetch method, but whenever I get a error from server component, I get this error message and can't get original error object anywhere.
```ts
console.error(
            `A query that was dehydrated as pending ended up rejecting. [${query.queryHash}]: ${error}; The error will be redacted in production builds`,
          )
```

I found [comments](https://github.com/TanStack/query/pull/7867#issuecomment-2365162267) about reason for redacted error (retry in client component), 
however there are also errors that do not need to be retried because they will failed anyway (like auth token expired)

So isn't there a way to handle error content before dehydrate query.promise?
Thank you!

**Comment by TkDodo** [comment]
I think you can dehydrate the query client, which just puts it into an object. Then, you can map over the `queries` and transform `promise` yourself. Something like:

```
queryClient.prefetchQuery({
  queryKey: ['posts'],
  queryFn: getPosts,
})

const dehydrated = dehydrate(queryClient)

dehydrated.queries = dehydrated.queries.map(query => ...)

<HydrationBoundary state={dehydrate(queryClient)}>
```

**Reply by yrnana** [reply]
@TkDodo thank you for fast answer!
I tried your answer, but I got same `redacted` message.
Maybe `catch` can't be removed from promise chain, there are no way to get original error object

```ts
dehydrated.queries = dehydrated.queries.map((query) => ({
  ...query,
  promise: query.promise?.catch((reason) => {
    console.log(reason); // redacted
    return Promise.reject(reason);
  }),
}));
```

**Reply by focux** [reply]
I'm experiencing the same error as well. I'm using prefetching on the server component without awaiting the response and then hydratating with an hydratation boundary and a suspense query. It also seems to break the server rendered page, even with an error boundary. What I'm trying to do is handling forbidden/not found errors from the server but it's been impossible to do it so far.

Any suggestion @TkDodo? I think I might be missing anything because this should be a common use-case. My code looks kinda like this:

```ts
export default function Page() {

 void queryClient.prefetchQuery(...);

return return (
    <HydrationBoundary state={dehydrate(queryClient)}>
      <PageContent />
    </HydrationBoundary>
  );
}
```

And then inside the `PageComponent` there's a suspense query.

**Reply by TkDodo** [reply]
does try/catch inside the `queryFn` with a re-throw work?

**Reply by stakbucks** [reply]
Nope. It's still caught [here](https://github.com/TanStack/query/blob/aef7ba209aeb53338dca262d48aef5719e8ac792/packages/query-core/src/hydration.ts#L87) and throws Error('redacted'). It makes it really hard to debug 😢

**Reply by focux** [reply]
@TkDodo Sorry, I've missed your answer before but ya, as @stakbucks pointed out, it doesn't work either and it makes handling error impossible

**Reply by TkDodo** [reply]
not really sure how to proceed here. The handling was added to avoid having server errors be sent to the client, as they might contain information that isn’t meant for the client.

@juliusmarminge @Ephem @phryneas what would your suggestions be ?

**Reply by focux** [reply]
I would have expected that you would catch and redact the errors that you don't want to show yourself.

In my case, I'm calling an API, and I would love to handle the status code errors returned by the API, especially now that next.js has functions to show especial pages depending on the error like `unauthenticated()‘, `forbidden()‘, etc.

Also, not sure if it has been others experience but the error boundary is not catching the error either.

**Reply by Ephem** [reply]
I meant to look at this last week but got sidetracked. I think there are several pains here. First, I think we'll want to make it configurable how to handle errors in promises being de/rehydrated in general. I think redacting is a good default, but you should be able to opt in to not redacting if you are using a framework like Next that does it for you. It should also be possible to silence the errors on the server side since it litters logs etc.

The second issue is you might want to do this on a query-level. So even if you are redacting errors by default, you might want to catch and transform the error into something that is not redacted for specific queries. Maybe even some way of avoiding retries, but I'm not sure we need to do something specific about that one. I think if you can transform the error yourself, you could use the function form of retries to accomplish this in userland.

I haven't gotten as far as figuring out the API-details for this yet though, but I'm very much dealing with the same pains on pretty much a daily basis, so it's at the top of my OSS-priority list. 😅

**Reply by stakbucks** [reply]
@focux @yrnana  For stopgap, I just patched the dehydrate code from [here](https://github.com/TanStack/query/blob/aef7ba209aeb53338dca262d48aef5719e8ac792/packages/query-core/src/hydration.ts). Which is literally the same except that it doesn't throw redact Error when query fails.


```typescript
function defaultTransformerFn(data: any): any {
  return data;
}

function dehydrateQuery(query: Query, serializeData: TransformerFn): DehydratedQuery {
  return {
    state: {
      ...query.state,
      ...(query.state.data !== undefined && {
        data: serializeData(query.state.data)
      })
    },
    queryKey: query.queryKey,
    queryHash: query.queryHash,
    ...(query.state.status === 'pending' && {
      promise: query.promise?.then(serializeData)
    }),
    ...(query.meta && { meta: query.meta })
  };
}

function dehydrateMutation(mutation: Mutation): DehydratedMutation {
  return {
    mutationKey: mutation.options.mutationKey,
    state: mutation.state,
    ...(mutation.options.scope && { scope: mutation.options.scope }),
    ...(mutation.meta && { meta: mutation.meta })
  };
}

export function dehydrate(client: QueryClient, options: DehydrateOptions = {}): DehydratedState {
  const filterMutation =
    options.shouldDehydrateMutation ??
    client.getDefaultOptions().dehydrate?.shouldDehydrateMutation ??
    defaultShouldDehydrateMutation;

  const mutations = client
    .getMutationCache()
    .getAll()
    .flatMap(mutation => (filterMutation(mutation) ? [dehydrateMutation(mutation)] : []));

  const filterQuery =
    options.shouldDehydrateQuery ??
    client.getDefaultOptions().dehydrate?.shouldDehydrateQuery ??
    defaultShouldDehydrateQuery;

  const serializeData =
    options.serializeData ??
    client.getDefaultOptions().dehydrate?.serializeData ??
    defaultTransformerFn;

  const queries = client
    .getQueryCache()
    .getAll()
    .flatMap(query => (filterQuery(query) ? [dehydrateQuery(query, serializeData)] : []));

  return { mutations, queries };
}


```

**Reply by focux** [reply]
@stakbucks Great! Will give it a try, thank you.

@Ephem @TkDodo Not sure if this is the right thread to ask, but another QoL improvement would be making the suspense query dehydrate/hydrate the headers that were sent on the initial/prefetched query.

Right now, if I prefetch a query on a server components, I need to pass down the headers (cookie headers mostly) to the component that is calling the `useSuspenseQuery`, otherwise, it throws an authorization error, I guess that is because of the SSR pass of the suspense query, although I would've expect that if the query was initiated on the server, all the headers would be also dehydrated/hydrated.

**Reply by phryneas** [reply]
In Apollo Client, we handle it like this:

* errors during SSR of client components will cause a rerender in the browser anyways -> we transport nothing, the query will be re-triggered in the browser, if it errors there, too, the user will see an error boundary.
* when transporting data that was fetched in RSC to prime a Client Component cache:
  * the special class of "graphql errors" will be transported over
  * any error in the link chain (e.g. when calling `fetch` will not be transported over. The query will be retried in the browser, if it errors there, too, the user will see an error boundary.


Generally, the concept of "will retry in the browser, if it errors there, too, the user will see an error boundary." is how React itself handles these kinds of errors, so we try to stay in line with that.
Only exception here are errors encoded within the query response, as these are meant to be sent to the browser anyways.
Of course, you don't have that specific "class" as you don't only support GraphQL, but maybe you could allow for a callback that lets the user decide if a specific error can be transported over or not?

**Reply by ghost** [reply]
I may be missing something in this [but Nextjs support a callback for server-side errors](https://nextjs.org/docs/app/api-reference/file-conventions/instrumentation#onrequesterror-optional) (since it will redact errors that are reaching the client as well)

Will the failed queries throw an error in a way that will hit that callback? I have the same setup with the author and I'm having the same issue with debugging these errors and trying to figure out how to proceed.

**Reply by TkDodo** [reply]
> Will the failed queries throw an error in a way that will hit that callback?

I’m not sure, can you try it out?

**Reply by ghost** [reply]
@TkDodo the failed queries will throw an error that will be caught by the `onRequestError` callback but they are still marked as "redacted". Which makes me think, if they've been caught server-side, why have they been redacted?

**Comment by edshav** [comment]
Hi, has anyone found out how to handle errors in this case?

**Comment by matheuspiresbbtv** [comment]
Hey everyone, any news on this case?

**Comment by spacerrsdreams** [comment]
Any news on this ?
